File: backend/app/services/plan_schedule.py
"""
Fälligkeiten eines Budgetplan-Eintrags.

Die Frage „in welchen Monaten eines Jahres fällt dieser Eintrag an, und wie
oft je Monat" wird an drei Stellen gebraucht: im Plan-Ist-Abgleich, in den
Monatssummen der Budgetplan-Ansicht und in den Fälligkeitswarnungen. Bisher
gab es sie nur im Frontend (`Budgetplan.tsx:getApplicableMonths`) — dort mit
einem Fehler bei `weekly`, das wie `monthly` behandelt wurde und deshalb
einfach statt 4,33-fach in die Summen einging.
"""

from __future__ import annotations

from datetime import date
from typing import List, Optional

#: Wie oft ein Eintrag pro Monat anfällt. 52/12 ≈ 4.33 Wochen je Monat —
#: derselbe Faktor, mit dem `budget_multimodal.py` bereits rechnet.
OCCURRENCES_PER_MONTH = {
    "weekly": 52 / 12,
    "monthly": 1.0,
    "quarterly": 1.0,
    "halfyearly": 1.0,
    "yearly": 1.0,
    "once": 1.0,
}

#: Monatsabstand zwischen zwei Fälligkeiten.
MONTH_STEP = {
    "weekly": 1,
    "monthly": 1,
    "quarterly": 3,
    "halfyearly": 6,
    "yearly": 12,
}


def occurrences_per_month(periodicity: Optional[str]) -> float:
    """Anzahl Fälligkeiten je Monat, in dem der Eintrag anfällt."""
    return OCCURRENCES_PER_MONTH.get(periodicity or "monthly", 1.0)


def applicable_months(
    periodicity: Optional[str],
    start_date: date,
    end_date: Optional[date],
    year: int,
) -> List[int]:
    """Monate (1–12) des Jahres, in denen der Eintrag anfällt.

    Der Ankermonat ist der Startmonat: ein quartalsweiser Eintrag ab März
    fällt im März, Juni, September und Dezember an, nicht im Januar.
    """
    if start_date.year > year:
        return []
    if end_date is not None and end_date.year < year:
        return []

    first = start_date.month if start_date.year == year else 1
    last = end_date.month if end_date is not None and end_date.year == year else 12
    if first > last:
        return []

    step = MONTH_STEP.get(periodicity or "monthly", 1)
    anchor = start_date.month

    if (periodicity or "monthly") == "once":
        return [anchor] if start_date.year == year and first <= anchor <= last else []

    return [m for m in range(first, last + 1) if (m - anchor) % step == 0]


def monthly_equivalent(amount: float, periodicity: Optional[str]) -> float:
    """Betrag auf einen Monatswert umgerechnet — für Vergleiche über Perioden."""
    p = periodicity or "monthly"
    if p == "weekly":
        return amount * OCCURRENCES_PER_MONTH["weekly"]
    step = MONTH_STEP.get(p, 1)
    return amount / step

File: frontend/src/lib/planSchedule.ts
/**
 * Fälligkeiten eines Budgetplan-Eintrags.
 *
 * Gegenstück zu `backend/app/services/plan_schedule.py`. Die Frage „in welchen
 * Monaten fällt dieser Eintrag an, und wie oft je Monat" stand vorher doppelt
 * im Frontend — in `Budgetplan.tsx` und in `Forecast.tsx` — und beide Kopien
 * behandelten `weekly` wie `monthly`, zählten einen wöchentlichen Eintrag also
 * einfach statt 4,33-fach.
 */

export interface PlanSchedule {
  periodicity: string;
  start_date: string;
  end_date: string | null;
}

/** Wie oft ein Eintrag je Monat anfällt, in dem er überhaupt anfällt. */
const OCCURRENCES_PER_MONTH: Record<string, number> = { weekly: 52 / 12 };

/** Monatsabstand zwischen zwei Fälligkeiten. */
const MONTH_STEP: Record<string, number> = {
  weekly: 1,
  monthly: 1,
  quarterly: 3,
  halfyearly: 6,
  yearly: 12,
};

export function occurrencesPerMonth(periodicity: string): number {
  return OCCURRENCES_PER_MONTH[periodicity] ?? 1;
}

/** Betrag einer Fälligkeit auf den Monatswert gerechnet. */
export function monthlyAmount(amount: number, periodicity: string): number {
  return amount * occurrencesPerMonth(periodicity);
}

/**
 * Monate (1–12) des Jahres, in denen der Eintrag anfällt.
 *
 * Ankermonat ist der Startmonat: ein quartalsweiser Eintrag ab März fällt im
 * März, Juni, September und Dezember an, nicht im Januar.
 */
export function applicableMonths(entry: PlanSchedule, year: number): number[] {
  const sd = new Date(entry.start_date + "T00:00:00");
  const ed = entry.end_date ? new Date(entry.end_date + "T00:00:00") : null;

  const startM =
    sd.getFullYear() < year ? 1 : sd.getFullYear() === year ? sd.getMonth() + 1 : null;
  if (startM === null) return [];

  const endM = ed
    ? ed.getFullYear() > year
      ? 12
      : ed.getFullYear() === year
        ? ed.getMonth() + 1
        : null
    : 12;
  if (endM === null) return [];

  const step = MONTH_STEP[entry.periodicity] ?? 1;
  const anchor = sd.getMonth() + 1;
  const months: number[] = [];
  for (let m = startM; m <= endM; m++) {
    if (((m - anchor) % step + step) % step === 0) months.push(m);
  }
  return months;
}
